Rethrow original failure in OkHttpClientTestRule

yschimke opened the pull request:
Related to https://github.com/square/okhttp/issues/6033

Want to avoid failures like

```
okhttp3.internal.ws.WebSocketHttpTest > serverMaxWindowBitsTooHigh FAILED
    java.lang.AssertionError: expected:<0> but was:<1>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:120)
        at org.junit.Assert.assertEquals(Assert.java:146)
        at okhttp3.OkHttpClientTestRule.ensureAllConnectionsReleased(OkHttpClientTestRule.kt:134)
        at okhttp3.OkHttpClientTestRule$apply$1.evaluate(OkHttpClientTestRule.kt:178)
        at okhttp3.testing.PlatformRule$apply$1.evaluate(PlatformRule.kt:69)
        at okhttp3.TestLogHandler$1.evaluate(TestLogHandler.java:50)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
```

swankjesse commented on okhttp-testing-support/src/main/kotlin/okhttp3/OkHttpClientTestRule.kt:
```diff
@@ -136,8 +136,13 @@ class OkHttpClientTestRule : TestRule {
   }
 
   private fun ensureAllTaskQueuesIdle() {
+    val entryTime = System.currentTimeMillis()
```
If you’re measuring durations, you gotta use nanoTime. In this case it doesn’t really matter but I prefer to do the proper thing everywhere.

swankjesse commented on okhttp-testing-support/src/main/kotlin/okhttp3/OkHttpClientTestRule.kt:
```diff
@@ -175,9 +182,25 @@ class OkHttpClientTestRule : TestRule {
-          ensureAllConnectionsReleased()
-          releaseClient()
-          ensureAllTaskQueuesIdle()
+          try {
```
try/catch and throw inside finally is icky; it’s really difficult control flow to follow.

Maybe something like this will work better?

```
val failure: Throwable? = null

try {
 // ...
} catch (t: Throwable) {
  if (failure != null) failure = t;
}

try {
 // cleanup 1
} catch (t: Throwable) {
  if (failure != null) failure = t;
}

try {
 // cleanup 2
} catch (t: Throwable) {
  if (failure != null) failure = t;
}

if (failure != null) throw t
```

swankjesse replied:
One other option is to make closeables for the three cleanup tasks (ensureAllConnectionsReleased(), releaseClient(), ensureAllTaskQueuesIdle()) and use `use {}` on each of ’em. That also permits suppressed exceptions!

yschimke replied:
In this case we want the test to fail if our test is causing these to break.  So we don't want to return a client that has active connections.  I'll rework this slightly, but the change as is should hopefully not hide the underlying problem.

I'm pretty sure this previously didn't hide the root cause, so not sure why it's regressed.

yschimke commented:
Testing with a deliberate failure

```
A
java.lang.AssertionError: A
	at org.junit.Assert.fail(Assert.java:89)
	at okhttp3.CallKotlinTest.testMockWebserverRequest(CallKotlinTest.kt:116)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at okhttp3.OkHttpClientTestRule$apply$1.evaluate(OkHttpClientTestRule.kt:170)
	at okhttp3.testing.PlatformRule$apply$1.evaluate(PlatformRule.kt:69)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:288)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:282)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
	Suppressed: java.lang.AssertionError: expected:<0> but was:<1>
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.failNotEquals(Assert.java:835)
		at org.junit.Assert.assertEquals(Assert.java:120)
		at org.junit.Assert.assertEquals(Assert.java:146)
		at okhttp3.OkHttpClientTestRule.ensureAllConnectionsReleased(OkHttpClientTestRule.kt:134)
		at okhttp3.OkHttpClientTestRule$apply$1.evaluate(OkHttpClientTestRule.kt:187)
		... 6 more
	Suppressed: java.lang.AssertionError: Queue still active after 1000 ms
		at org.junit.Assert.fail(Assert.java:89)
		at okhttp3.OkHttpClientTestRule.ensureAllTaskQueuesIdle(OkHttpClientTestRule.kt:147)
		at okhttp3.OkHttpClientTestRule.access$ensureAllTaskQueuesIdle(OkHttpClientTestRule.kt:40)
		at okhttp3.OkHttpClientTestRule$apply$1.evaluate(OkHttpClientTestRule.kt:200)
		... 6 more
```

yschimke commented:
Still failing

```
okhttp3.internal.ws.WebSocketHttpTest > overflowOutgoingQueue FAILED
    java.lang.AssertionError: expected:<0> but was:<1>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:120)
        at org.junit.Assert.assertEquals(Assert.java:146)
        at okhttp3.OkHttpClientTestRule.ensureAllConnectionsReleased(OkHttpClientTestRule.kt:134)
        at okhttp3.OkHttpClientTestRule$apply$1.evaluate(OkHttpClientTestRule.kt:187)
        at okhttp3.testing.PlatformRule$apply$1.evaluate(PlatformRule.kt:69)
        at okhttp3.TestLogHandler$1.evaluate(TestLogHandler.java:50)
```

https://app.circleci.com/pipelines/github/square/okhttp/3216/workflows/76f56479-cadc-4202-bb5e-5745c0a2a175/jobs/13505/steps

I think that is enough to conclude the test is passing, but that cleanup isn't immediate.

yschimke commented:
@swankjesse apologies, going to land this without final approval to try to get some signal from any failing tests etc.  But I've hopefully addressed your review points. It is only test classes, so seems minimal risk.

yschimke commented:
Failure in CI on master

https://app.circleci.com/pipelines/github/square/okhttp/3219/workflows/8563ba9a-628d-43a4-a240-7e21be452d90/jobs/13521/tests

```
responseCachingByResponseCode - okhttp3.CacheTest

java.io.IOException: Gave up waiting for queue to shut down
	at okhttp3.mockwebserver.MockWebServer.shutdown(MockWebServer.kt:453)
	at okhttp3.CacheTest.assertCached(CacheTest.java:202)
	at okhttp3.CacheTest.responseCachingByResponseCode(CacheTest.java:137)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at okhttp3.OkHttpClientTestRule$apply$1.evaluate(OkHttpClientTestRule.kt:172)
	at okhttp3.internal.io.InMemoryFileSystem$apply$1.evaluate(InMemoryFileSystem.kt:41)
	at okhttp3.testing.PlatformRule$apply$1.evaluate(PlatformRule.kt:69)
```

yschimke commented:
> LGTM
> 
> Sorry for holding you up! I definitely prefer unblocking over gatekeeping.

pffftttt.... Don't worry about that. Nothing here is urgent, just wanted to flag since you may have wanted a second pass when it was convenient. And if I call it out then it's clear that I'm not just landing things without a review as standard.
